overlord/devicestate/users.go

```diff
@@ -0,0 +1,359 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package devicestate
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os/user"
+	"path/filepath"
+	"time"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/assertstate"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/strutil"
+)
+
+var (
+	osutilAddUser = osutil.AddUser
+	osutilDelUser = osutil.DelUser
+	userLookup    = user.Lookup
+)
+
+// UserError is used for errors which are known to snapd. This is errors
+// originate from this code here.
+type UserError struct {
+	Err error
+}
+
+func (e *UserError) Error() string {
+	return e.Err.Error()
+}
+
+// CreatedUser is the returned result which wraps some resulting data
+// from a create user operation.
+type CreatedUser struct {
+	Username string
+	SSHKeys  []string
+}
+
+// CreateUser creates linux user based on the passed email, username
+// and public ssh keys for the created user account are determined
+// from Ubuntu store based on the passed email.
+func CreateUser(st *state.State, sudoer bool, email string) (CreatedUser, error) {
+	if email == "" {
+		return CreatedUser{}, &UserError{Err: fmt.Errorf("cannot create user: 'email' field is empty")}
+	}
+
+	db := assertstate.DB(st)
+	u := &createUserOpts{
+		assertDb: db,
+		modelAs:  nil,
+		serialAs: nil,
+		isSudoer: sudoer,
+	}
+
+	storeService := snapstate.Store(st, nil)
+	st.Unlock()
+	username, opts, err := getUserDetailsFromStore(storeService, email)
+	st.Lock()
+
+	if err != nil {
+		return CreatedUser{}, &UserError{Err: err}
+	}
+	opts.Sudoer = sudoer
+	createdUser, err := u.createUser(st, opts, username, email)
+	if err != nil {
+		return CreatedUser{}, err
+	}
+	return createdUser, nil
+}
+
+// CreateKnownUsers create known user(s)
+// user details are fetched from existing system user assertions.
+// If no email is passed, all known users will be created based on valid system user assertions.
+// If email is passed, only corresponding system user assertion is used for user creation.
+func CreateKnownUsers(st *state.State, mgr *DeviceManager, sudoer bool, email string) ([]CreatedUser, error) {
```

Comment:
it doesn't make sense for this to take the DeviceManager now nor it needs to now. It can use the findModel and findSerial helpers in this package.

Reply:
Really nice catch, makes it so much cleaner!